map/extrapolation/extrapolator.cpp

```diff
@@ -0,0 +1,172 @@
+#include "map/extrapolation/extrapolator.hpp"
+
+#include "geometry/distance_on_sphere.hpp"
+
+#include "platform/platform.hpp"
+
+#include "base/logging.hpp"
+
+#include <chrono>
+#include <memory>
+
+namespace
+{
+uint64_t constexpr kMaxExtrapolationTimeMs = 1000;
+uint64_t constexpr kExtrapolationPeriodMs = 200;
+double constexpr kMaxExtrapolationSpeedMPS = 120.0;
+
+double Delta(double param1, double param2, uint64_t timeBetweenPointsMs, uint64_t timeAfterParam2Ms)
+{
+  return (param2 - param1) * timeAfterParam2Ms / timeBetweenPointsMs;
+}
+
+class LinearExtrapolator
+{
+public:
+  LinearExtrapolator(uint64_t timeBetweenMs, uint64_t timeAfterMs)
+    : m_timeBetweenMs(timeBetweenMs), m_timeAfterMs(timeAfterMs)
+  {
+  }
+
+  double Extrapolate(double x1, double x2)
+  {
+    return x2 + (x2 - x1) / m_timeBetweenMs * m_timeAfterMs;
+  }
+
+private:
+  uint64_t m_timeBetweenMs;
+  uint64_t m_timeAfterMs;
+};
+}  // namespace
+
+namespace extrapolation
+{
+using namespace std;
+
+location::GpsInfo LinearExtrapolation(location::GpsInfo const & gpsInfo1,
+                                      location::GpsInfo const & gpsInfo2,
+                                      uint64_t timeAfterPoint2Ms)
+{
+  auto const timeBetweenPointsMs =
+      static_cast<uint64_t>((gpsInfo2.m_timestamp - gpsInfo1.m_timestamp) * 1000);
+
+  location::GpsInfo extrapolated = gpsInfo2;
+  LinearExtrapolator extrapolator(timeBetweenPointsMs, timeAfterPoint2Ms);
+
+  extrapolated.m_timestamp += timeAfterPoint2Ms;
+
+  extrapolated.m_longitude = extrapolator.Extrapolate(gpsInfo1.m_longitude, gpsInfo2.m_longitude);
+
+  extrapolated.m_latitude = extrapolator.Extrapolate(gpsInfo1.m_latitude, gpsInfo2.m_latitude);
+
+  extrapolated.m_horizontalAccuracy =
+      extrapolator.Extrapolate(gpsInfo1.m_horizontalAccuracy, gpsInfo2.m_horizontalAccuracy);
+
+  extrapolated.m_altitude = extrapolator.Extrapolate(gpsInfo1.m_altitude, gpsInfo2.m_altitude);
+
+  if (gpsInfo1.HasVerticalAccuracy() && gpsInfo2.HasVerticalAccuracy())
+  {
+    extrapolated.m_verticalAccuracy =
+        extrapolator.Extrapolate(gpsInfo1.m_verticalAccuracy, gpsInfo2.m_verticalAccuracy);
+  }
+
+  if (gpsInfo1.HasBearing() && gpsInfo2.HasBearing())
+    extrapolated.m_bearing = extrapolator.Extrapolate(gpsInfo1.m_bearing, gpsInfo2.m_bearing);
+
+  if (gpsInfo1.HasSpeed() && gpsInfo2.HasSpeed())
+    extrapolated.m_speed = extrapolator.Extrapolate(gpsInfo1.m_speed, gpsInfo2.m_speed);
```

Comment:
Again, why not to use, at least, momentum methods, or exponential decay, at least? IMHO, this degree 1 has quite low quality. Have you measured it somehow?

Reply:
According to benchmark linear method give the following mathematical expectation and dispersion/standard deviation  for deviations extrapolated points from track.
```
    /Users/vladimirbykoyanko/src_github/omim/cmake-build-debug/extrapolation_benchmark -csv_path=/Users/vladimirbykoyanko/tracks/users/trafin_log.20180517-0000.track.csv
LOG TID(1) INFO   4.6257e-05 extrapolation_benchmark/extrapolation_benchmark.cpp:228 main() General tracks statistics.
  Number of tracks: 886 
  Number of track points: 567362 
  Average points per track: 640 
  Average track length: 15325.8 meters
LOG TID(1) INFO      115.825 extrapolation_benchmark/extrapolation_benchmark.cpp:319 main() 
  Processed 153443 points.
    767215 extrapolations is calculated.
   Projection is calculated for 767215 extrapolations.
LOG TID(1) INFO      115.825 extrapolation_benchmark/extrapolation_benchmark.cpp:321 main() Expected value for each extrapolation:
LOG TID(1) INFO      115.825 extrapolation_benchmark/extrapolation_benchmark.cpp:327 main() Extrapolation 1 , 200 seconds after point two. Expected value = 0.705745 meters. Variance = 1.29414 . Standard deviation = 1.1376
LOG TID(1) INFO      115.825 extrapolation_benchmark/extrapolation_benchmark.cpp:327 main() Extrapolation 2 , 400 seconds after point two. Expected value = 1.39173 meters. Variance = 4.76746 . Standard deviation = 2.18345
LOG TID(1) INFO      115.825 extrapolation_benchmark/extrapolation_benchmark.cpp:327 main() Extrapolation 3 , 600 seconds after point two. Expected value = 2.04523 meters. Variance = 9.80405 . Standard deviation = 3.13114
LOG TID(1) INFO      115.825 extrapolation_benchmark/extrapolation_benchmark.cpp:327 main() Extrapolation 4 , 800 seconds after point two. Expected value = 2.63355 meters. Variance = 15.9907 . Standard deviation = 3.99883
LOG TID(1) INFO      115.825 extrapolation_benchmark/extrapolation_benchmark.cpp:327 main() Extrapolation 5 , 1000 seconds after point two. Expected value = 3.13071 meters. Variance = 22.8244 . Standard deviation = 4.77749
```
So for the worst case we have:
Mathematical expectation: 3.1 meters
Standard deviation = 4.8 meters

For me it looks good enough. Please take into account that this extrapolation is used only in following mode in car and bicycle navigation. So with route matching with passage 100 and 60 meters.